[Feature] 로그인 기능 구현 - 기본 로그인

dev-msj opened the pull request:
## 작업 내용

difinition.md 문서 작성
java-jwt 라이브러리 추가
JwtService 구현
UserAuthMapper 구현
UserAuthService 구현
TokenInterceptor 구현
UserController 구현
각 component들에 필요한 model들 구현

## 이슈 번호

#23 Login API 구현

dev-msj committed:
docs : TODO 정의

dev-msj committed:
chore : java jwt 라이브러리 추가

dev-msj committed:
feat : TokenInterceptor 생성 및 등록

dev-msj committed:
feat : UserLoginRequest, UserAuth model 추가

dev-msj committed:
feat : UserAuthMapper 구현

dev-msj committed:
feat(JwtExpiredState) : Token 만료 상태 정의

dev-msj committed:
feat(JwtProperty) : jwt 공통 Payload 정의

추후 Valt로 분리 예정

dev-msj committed:
feat(JwtToken) : access/refresh token을 담는 모델 생성

dev-msj committed:
refactor(UserLoginRequest) : Serializable 상속

dev-msj committed:
refactor(UserAuthMapper) : token 인증과 관련된 쿼리만 하도록 변경

uid에 해당하는 UserAuth가 존재하는지 확인하는 쿼리 추가
uid와 refresh token 해당하는 UserAuth가 존재하는지 확인하는 쿼리 추가
refresh token 업데이트 쿼치 추가

dev-msj committed:
style(JwtService) : Jwt 관리 로직 설계

dev-msj committed:
style(TokenInterceptor) : 토큰 인터셉터 처리 로직 설계

dev-msj committed:
style(UserAuthService) : 로그인 로직 설계

dev-msj committed:
feat(UserAuthMapper) : uid에 해당하는 UserAuth를 가져오는 쿼리 추가

dev-msj committed:
feat(UserAuthContorller) : request mapping 추가 - login, refresh token

dev-msj committed:
fix : model과 mapper에서 access token 삭제

UserAuth 테이블에서 access token을 다루지 않으므로 삭제한다.

dev-msj committed:
feat : AuthorizationException 생성

dev-msj committed:
refactor(JwtToken) : AllArgsConstructor 어노테이션 추가

dev-msj committed:
feat : JwtService 구현

dev-msj committed:
style(JwtService) : public 메서드가 private 메서드보다 위에 있도록 변경

dev-msj committed:
style(JwtService) : 메서드 이름 변경

dev-msj committed:
feat : UserAuthService 구현

dev-msj committed:
feat : TokenInterceptor 구현

dev-msj committed:
refactor(JwtService) :  checkTokenInfoExistInDB 메서드의 return 값 반전

dev-msj committed:
style : 주석 및 로깅에 필요한 콘솔 아웃 추가

dev-msj committed:
refractor(TokenInterceptor) : response status code를 ErrorCode에서 받도록 변경

dev-msj committed:
fix(UserAuthService) : salt를 UserAuth에서 가져와서 사용하도록 수정

dev-msj committed:
refactor(InterceptorConfig) : refresh 요청은 interceptor 제외

dev-msj committed:
refactor : api prefix 적용

InterceptorConfig, UserAuthController의 request mapping 주소에 api prefix 적용

dev-msj committed:
style(UserAuthService) : 사용하지 않는 import 제거

dev-msj committed:
style : 사용하지 않는 class 삭제

jinspark-lab commented on src/main/java/com/flab/oasis/constant/JwtProperty.java:
```diff
@@ -0,0 +1,10 @@
+// 추후 Valt로 분리 예정
+public class JwtProperty {
+    public static final String SECRET_KEY = "Oasis Auth";
+    public static final String ISSUER = "Oasis Server";
```
환경에 따라 이슈가 될 수 있습니다. 띄어쓰기는 가급적 사용하지마세요.

dev-msj replied:
네 반영했습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/service/JwtService.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    public boolean checkTokenInfoExistInDB(DecodedJWT decodedJWT) {
+        if (decodedJWT.getSubject().equals(ACCESS_TOKEN)) {
```
로직이 어떤 의도인지 잘 모르겠습니다.

dev-msj replied:
access token이면 token의 id,
refresh token이면 token의 id와 refresh
가 DB에 존재하는지 확인하는 로직입니다.

쿼리에서 Exists가 누락되어 수정했습니다.

jinspark-lab replied:
그걸 여쭤보는게 아닙니다..

dev-msj replied:
Access Token은 토큰이 유효하면 요청한 정보를 넘겨주는 것이 목적인데,
목적에 맞지 않는 불필요한 조건절을 작성했었네요...
해당 로직은 삭제했습니다.

jinspark-lab commented on src/main/resources/mapper/UserAuthMapper.xml:
```diff
@@ -0,0 +1,46 @@
+               REFRESH_TOKEN AS refreshToken
+            FROM USER_AUTH
+            WHERE UID = #{uid}) AS result
+        FROM DUAL
```
Dual 은 왜 쓰시는건가요?

dev-msj replied:
원래는 `Select Exists (sub query) From Dual`이어야 하는데, Exists가 누락되었네요.
다시 수정해서 올렸습니다.

dev-msj replied:
Dual은 삭제 했습니다.
MySQL에서는 더미 테이블 없어도 Select가 된다는 것을 몰랐네요...

jinspark-lab reviewed (changes requested):
로그인 로직 및 토큰 처리가 이상합니다. 테스트는 되셨나요? 아마 프론트엔드가 없어서 테스트가 잘 안되신것 같긴합니다만.. 로직 전반에 대해 좀 더 공부해보시고 수정해보세요.

dev-msj committed:
style(JwtProperty) : 띄어쓰기 삭제

dev-msj committed:
fix(UserAuthMapper) : 누락된 Operator 추가

dev-msj committed:
style(JwtService) : public 메서드가 private 메서드보다 상단에 위치하도록 변경

dev-msj committed:
test : JwtServiceTest 작성

jinspark-lab commented on src/test/java/com/flab/oasis/service/JwtServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+                .withExpiresAt(new Date(issueDate.getTime() + 1))
+                .sign(algorithm);
+
+        Thread.sleep(1);
```
Sleep 은 의도가 뭔가요

dev-msj replied:
원래는 토큰을 expire 시키기 위해서 사용했는데, issuedAt과 ExpiresAt을 동일하게 줘서 생성과 동시에 expire 시킬 수 있었네요...
테스트 작성할 때, 이런 무의미한 구문이 작성되지 않도록 조금 더 신중하도록 하겠습니다

jinspark-lab reviewed (commented):
관련해서 수업시간에 피드백드리겠습니다.

dev-msj committed:
fix(JwtService) testDecodeJwtThrowExpired에서 무의미한 sleep 제거

dev-msj committed:
fix(UserAuthMapper) : 불필요한 dual 테이블 제거

dev-msj committed:
test(JwtServiceTest) : testDecodeJwtThrowExpired에서 토큰이 생성과 동시에 expire되도록 변경

dev-msj committed:
refactor(JwtService) : Access Token의 id가 db에 있는지 확인하는 로직 삭제

dev-msj committed:
style(JwtService) : 메서드 decodeJWT를 verifyJwt로 변경

dev-msj committed:
fix(JwtService) : 토큰 발행시 jit 대신 private claim을 사용하도록 변경

dev-msj committed:
refactor(TokenInterceptor) : Access Token의 id가 db에 있는지 확인하는 로직 삭제

dev-msj committed:
refactor(UserAuthController) : access/refresh token을 헤더/cookie에 담아 반환하도록 변경

dev-msj committed:
style(JwtServiceTest) : 패키지 경로 변경

jinspark-lab reviewed (commented):
이 부분이 잘 안잡히네요.. 전반적으로 코드 리뷰가 지적한 부분만 좋아지는 부분이 아직 계속 문제입니다. 
서두르시는 것 같은데, 빨리 머지시킬 생각은 일단 접어두시고, 먼저 내용을 충분히 이해하시고 코드를 작성하셔야합니다... 저도 마음같아서는 다 잡아드리고 싶은데 이 부분을 이번주에 마지막으로 논의해보시져

jinspark-lab commented on src/main/java/com/flab/oasis/controller/UserAuthController.java:
```diff
@@ -0,0 +1,52 @@
+            @CookieValue("refreshToken") String refreshToken, HttpServletResponse response) {
+        JwtToken jwtToken = jwtService.reissueJwtToken(refreshToken);
+
+        response.setHeader("Authorization", String.format("%s %s", "Bearer", jwtToken.getAccessToken()));
```
흠...

dev-msj committed:
refactor(JwtService) : 토큰 발급 시 uid가 subject claim으로 사용되도록 변경

dev-msj committed:
refactor(UserAuthController) : 토큰 반환 방식 변경

access token을 cookie에 담고 refresh token을 header에 담도록 변경

dev-msj committed:
style(UserAuthMapper) : 사용되지 않는 쿼리 삭제

dev-msj committed:
refactor : UserAuth 테이블에서 refresh token 삭제

dev-msj committed:
feat(RedisConfig) : defaultTemplate 추가

dev-msj committed:
style(JwtProperty) : 토큰 만료 시간 변경

dev-msj committed:
refactor(AuthorizationException) : 에러 코드를 같이 받도록 변경

dev-msj committed:
feat(UserAuthRepository) : redis에서 refresh token을 관리하는 기능 추가

dev-msj committed:
refactor(JwtService) : redis에서 refresh token을 관리하도록 변경

dev-msj committed:
style(UserAuthService) : login 메서드명을 역할과 input/output이 드러나도록 변경

dev-msj committed:
refactor : cors 설정 추가 및 InterceptorConfig와 병합

dev-msj committed:
refactor(UserAuthController) : refresh token을 쿠키에 담도록 변경

dev-msj committed:
refactor(TokenInterceptor) : 토큰이 인증되면 컨트롤러로 넘기도록 변경

dev-msj committed:
test(JwtServiceTest) : 테스트 코드 재작성 및 목적과 맞지 않는 테스트 삭제

dev-msj committed:
style(UserLoginRequest) : 생성자 관련 어노테이션 추가

dev-msj committed:
test(UserAuthControllerTest) : 컨트롤러 테스트 추가

dev-msj committed:
feat(ExceptionController) : Custom Exception 메세지 response 기능 추가

dev-msj committed:
style(UserLoginRequest) : ToString 어노테이션 추가

dev-msj committed:
refactor : Log 메세지 format 적용

dev-msj committed:
refactor(TokenInterceptor) : access token expire 예외처리 추가

dev-msj committed:
docs(JwtService) : 주석 오류 수정

jinspark-lab commented on src/main/java/com/flab/oasis/config/WebConfig.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public void addCorsMappings(CorsRegistry registry) {
+        registry.addMapping("/**")
+                .allowedOrigins("http://localhost:3000")
```
좋은 습관이시기는 합니다만, 그냥 * 로 하셔도 됩니다. 어차피 localhost 만 등록하는건 의미도 없구요. 원래는 원격지 서버 주소를 입력해야하긴 합니다만 토이 프로젝트이니 전부 뚫어주셔요.

dev-msj replied:
네 반영했습니다.

jinspark-lab replied:
이거 다시 롤백하셨던데, 이유가 있으신가요? CORS 에서 * 가 적용 안되는건 말이 안되는데요. 이거 로컬호스트만 뚫으시면 나중에 클라이언트 대충이라도 만드셨을때 서비스는 안되십니다. 너무 오래 끌 수는 없으니 어떤 문제신지 설명 부탁드려요.

jinspark-lab commented on src/main/java/com/flab/oasis/controller/UserAuthController.java:
```diff
@@ -0,0 +1,58 @@
+        JwtToken jwtToken = jwtService.reissueJwtToken(refreshToken);
+
+        response.setHeader(AUTHORIZATION_HEADER, makeAuthorizationValue(jwtToken.getAccessToken()));
+        response.setHeader(SET_COOKIE, createCookie(jwtToken.getRefreshToken()));
```
지난 시간 피드백 똑같이 드리겠습니다. 마찬가지로 의도가 궁금합니다.

dev-msj replied:
2개의 토큰이 역할이 다르기 때문에 서로 분리된 형태로 관리해야겠다라고 생각했습니다.
그래서 비교적 안전하게 관리할 수 있는 쿠키에 refresh token을 넣고, access token을 웹 스토리지에서 관리하려고 해서 위처럼 작성했었습니다.
그런데 굳이 분리해서 관리할 필요가 없고, 오히려 access token도 쿠키에서 같이 관리하는 것이 더 좋을 것 같다라고 판단되어 토큰 모두 쿠키로 처리되도록 변경 후 반영했습니다!

jinspark-lab commented on src/main/java/com/flab/oasis/interceptor/TokenInterceptor.java:
```diff
@@ -0,0 +1,88 @@
+                JwtToken jwtToken = jwtService.reissueJwtToken(refreshToken);
+
+                response.setHeader("Authorization", String.format("%s %s", "Bearer", jwtToken.getAccessToken()));
+                response.setHeader("Set-Cookie", createCookie(jwtToken.getRefreshToken()));
```
Controller 와 Interceptor 모두에서 해주시는 이유가 있나요?

dev-msj replied:
프론트 로직에 대해 정확히 모르다보니 상황에 따라 refresh token을 직접 들고 토큰 재발급 요청을 하는 경우도 있지 않을까라는 생각에 컨틀롤러에도 해당 내용을 넣어놨었습니다.
그런데 로그인 화면 구현쪽도 같이 진행하다보니 리액트에서 인터셉터를 활용하여 요청 에러 시에 대한 핸들링이 가능하다는 것을 알게 되어서, 컨트롤러에서 해당 메서드는 삭제했습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/model/exception/ResponseError.java:
```diff
@@ -0,0 +1,16 @@
+public class ResponseError implements Serializable {
+    private static final long serialVersionUID = 1772002427691231525L;
+
+    private String message;
```
사소한 거긴 합니다만, StatueCode 에 의존하는 것보다 errorcode 를 명시적으로 전달해주는게 더 좋습니다. 이 피드백은 일단 무시해주시고,  관련해서는 수업 시간에 말씀드릴게요.

dev-msj replied:
네 알겠습니다!

jinspark-lab commented on src/main/java/com/flab/oasis/repository/UserAuthRepository.java:
```diff
@@ -0,0 +1,32 @@
+
+@Repository
+@RequiredArgsConstructor
+public class UserAuthRepository {
```
동작이 Service 에 더 가까워보입니다.

dev-msj replied:
Redis에서 Jwt 관련 로직을 처리하니 분리시켜야 된다고 생각해서 Repository를 따로 만들었는데,
이미 redisTemplate에 있는 내용들을 굳이 repository를 한번더 거쳐서 사용하고 있었네요...
해당 내용은 삭제하고 jwtService에서 redisTemplate를 직접 사용하도록 변경했습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/service/JwtService.java:
```diff
@@ -0,0 +1,99 @@
+            DecodedJWT decodedJWT = verifyJwt(refreshToken);
+
+            // refresh token이 redis에 존재하지 않으면 다시 로그인을 하도록 한다.
+            if (!userAuthRepository.existUserRefreshToken(decodedJWT.getSubject(), refreshToken)) {
```
DB 를 안 거치는 것 같은데 무슨 의도신가요?

dev-msj replied:
refresh token이 만료될 경우에도 DB에는 토큰이 계속 남아있게 될텐데, 이 리소스를 계속 가지고 있는 것이 과연 옳을까에 대해 의문이 들었습니다. 그래서 redis만 사용하더라도 문제는 없을까 생각해봤는데, access token이 존재하기 때문에 redis에 문제가 생기더라도 사용자가 로그인을 더 많이 해야할 뿐 이용에는 문제가 없을 것이다라고 판단이 되어서 redis만 사용하도록 변경했습니다.

그런데 답변을 작성하다보니 드는 생각이 사용자가 로그인을 더 많이 해야한다는 것 자체가 문제가 되네요...
그래서 redis에 문제가 생겨도 DB에서 가져와서 인증을 해야할 필요가 있는 것이었네요.
해당 부분은 재작성하도록 하겠습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/utils/LogUtils.java:
```diff
@@ -0,0 +1,27 @@
+
+import com.flab.oasis.constant.ErrorCode;
+
+public class LogUtils {
```
별도로 빼실거면 차라리 여기에서 로그를 찍는게 나을 것 같습니다.

jinspark-lab reviewed (commented):
테스트쪽 포함해서 몇몇 부분은 많이 개선됐네요. 수업시간 피드백 포함 몇가지 커멘트 드렸습니다.

dev-msj committed:
refactor : 토큰 발급을 쿠키로만 하도록 변경

dev-msj committed:
refactor(UserAuthController) : refresh token으로 토큰 재발급을 Interceptor에서만 하도록 변경

dev-msj committed:
refactor(JwtService) : UserAuthRepository의 내용을 JwtService로 이동

dev-msj committed:
refactor(WebConfig) : allowedOrigins를 모두 허용하도록 변경

추후 ncp에 배포하면 주소 반영할 예정

jinspark-lab commented on src/main/java/com/flab/oasis/interceptor/TokenInterceptor.java:
```diff
@@ -0,0 +1,94 @@
+                ).getValue();
+
+        try {
+            jwtService.verifyJwt(accessToken);
```
accessToken 에 대한 검증이 필요합니다. 최소한이라도 구현해주시는게 좋습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/service/JwtService.java:
```diff
@@ -0,0 +1,104 @@
+            throws TokenExpiredException, SignatureVerificationException, InvalidClaimException {
+        Algorithm algorithm = Algorithm.HMAC256(JwtProperty.SECRET_KEY);
+        JWTVerifier verifier = JWT.require(algorithm)
+                .withIssuer(JwtProperty.ISSUER)
```
최소한 uid 까지라도 검사가 필요합니다

dev-msj committed:
fix(WebConfig) : *로 cors 적용이 안되서 다시 기존 url 입력

dev-msj committed:
refactor : DB에서 Refresh Token을 관리하도록 변경

dev-msj committed:
refactor(LogUtils) : slf4j로 로그 메세지를 남기도록 변경

dev-msj committed:
refactor : LogUtils 변경사항 적용

dev-msj committed:
refactor(UserAuthService) : UserAuthMapper 관련된 처리는 UserAuthService에서 관리되도록 변경

dev-msj committed:
refactor(ErrorCode) : Service Unavailable 코드 추가

dev-msj committed:
feat : 인증 정보를 담을 UserSession 모델 추가

dev-msj committed:
refactor(UserAuthMapper) : parameter를 UserSession으로 받도록 변경

dev-msj committed:
feat : 유저 인증 정보를 관리하는 UserAuthRepository 구현

dev-msj committed:
refactor(UserAuthService) : 유저 인증 정보를 UserAuthRepository를 통해 가져오도록 변경

dev-msj committed:
refactor(UserAuth) : UserAuth 모델을 dao로 활용하도록 변경

password 비교 메서드 추가
UserSession 파싱 메서드 추가

dev-msj committed:
refactor(UserAuthMapper) : UserAuth 변경사항 반영

dev-msj committed:
refactor(UserAuthRepository) : UserSession 파싱을 UserAuth에서 하도록 변경

dev-msj committed:
refactor(UserAuthService) : password 비교를 UserAuth에서 하도록 변경

dev-msj committed:
refactor(JwtService) : DAO 모델명에는 DAO를 명시하도록 변경

dev-msj committed:
fix : DAO 패턴 삭제

dev-msj committed:
fix(UserAuthRepository) : DAO 삭제 관련 누락된 부분 반영

dev-msj committed:
refactor : UserSession 형태로 DB에서 가져오도록 변경

dev-msj committed:
refactor(JwtService) : Jwt 관리 로직 수정

access token과 refresh token 인증 로직 분리
private claim 추가 - uid
token 인증 시 private claim값을 확인

dev-msj committed:
refactor(TokenInterceptor) : access token과 refresh token 모두 인증하도록 수정

dev-msj committed:
style(JwtService) : refresh token 매칭 exception 메세지 수정

dev-msj committed:
fix(JwtService) : claim 값 형 변환 오류 수정

dev-msj committed:
fix(LogUtils) : 줄 바꿈 적용되도록 수정

dev-msj committed:
test(JwtServiceTest) : 로직 변경 내용에 맞게 테스트 수정

jinspark-lab reviewed (commented):
CORS 만 처리하면 머지해도 될 것 같습니다.

dev-msj committed:
refactor(WebConfig) : allow origin 모든 url 허용

jinspark-lab commented:
머지하겠습니다. 구글 로그인 적용 시 나머지 부분 좀 더 다듬으면 괜찮을 것 같습니다.
